### ISSUE TITLE
Explore a common design pattern for basis functions of spatial discretizations

### ISSUE THREAD

**Issue by Naktakala** [chi-tech/chi-tech#192]
For a given spatial discretization type, nodes are allocated on/in a cell based on the type of basis-functions used. The basis functions and their associated mappings to and from a reference cell is controlled via a `CellMapping` class.

We should look into the utilities afforded to a base-class `CellMapping` that will provide information like:
- Cell number of nodes
- Cell-face number of nodes
- Cell node locations
- Cell-face node locations
- Mapping a simulation-domain point onto reference coordinates-system point
- Mapping a reference coordinates-system point ont a simulation-domain point
- etc.

Basically we need more functionality from the base `CellMapping` class so that we can define more diverse spatial discretizations.

**Comment by Naktakala** [comment]
I think I got this down now.
